### PR TITLE
aks advanced network and nodepools config options

### PR DESCRIPTION
**Pull request by manoatsas**
Closing #43 and #75

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -74,7 +74,7 @@ data "tls_public_key" "public_key" {
 locals {
   # Network ip ranges
   vnet_cidr_block          = "192.168.0.0/16"
-  aks_subnet_cidr_block    = "192.168.1.0/24"
+  aks_subnet_cidr_block    = "192.168.16.0/20"
```
Comment (thpang) [inline main.tf:77]: This value should be adjusted based on the user's selection. If the network is `kubenet` then this should keep its old value. However, if it's `azure` then it should be assigned this value. Also what is driving this actual CIDR range: 192.168.16.0./20 why not 192.168.4.0/20 Just wondering why the large leap? This does not collide with any of the previous /24 items.

File: docs/CONFIG-VARS.md [hunk 2]
```diff
@@ -100,6 +100,11 @@ Additional node pools can be created separate from the default nodepool. This is
 | os_disk_size | Disk size for nodepool VMs in GB | number | |
 | min_nodes | Minimum number of nodes for the nodepool | number | Value must be between 0 and 100. Setting min and max node counts the same disables autoscaling |
 | max_nodes | Maximum number of nodes for the nodepool | number | Value must be between 0 and 100. Setting min and max node counts the same disables autoscaling |
+| max_pods | Maximum number of pods per node | number | Default is 110
+| mode |  Nodepool mode | string | System or User
+| enable_node_public_ip | Enable public IP for each node | bool | |
```
Comment (thpang) [inline docs/CONFIG-VARS.md:105]: So not sure why all of these items are being surfaced now: `mode`, `enable_node_public_ip`, `priority`, and `eviction_policy`. I am thinking most of these would be nice to have, but these are not part of this pr. If we need them for this change, please add more details in the pr description. If they are not needed, then they should be removed and placed in another pr.

File: modules/aks_node_pool/main.tf [hunk 3]
```diff
@@ -9,12 +9,21 @@ resource "azurerm_kubernetes_cluster_node_pool" "autoscale_node_pool" {
   proximity_placement_group_id = var.proximity_placement_group_id == "" ? null : var.proximity_placement_group_id
   vm_size                      = var.machine_type
   os_disk_size_gb              = var.os_disk_size
+  # TODO: enable after azurerm v2.37.0
```
Comment (thpang) [inline modules/aks_node_pool/main.tf:12]: Azurerm 2.38.0 has been out for about a month now. Is the need for `os_disk_type` tied to this if so lets make the update.

File: modules/aks_node_pool/main.tf [hunk 4]
```diff
@@ -31,11 +40,20 @@ resource "azurerm_kubernetes_cluster_node_pool" "static_node_pool" {
...
```
Comment (thpang) [inline modules/aks_node_pool/main.tf:43]: Same comment as above.

File: modules/aks_node_pool/variables.tf [hunk 5]
```diff
@@ -25,6 +25,25 @@ variable "os_disk_size" {
   default = 100
 }
 
+# TODO: enable after azurerm v2.37.0
```
Comment (thpang) [inline modules/aks_node_pool/variables.tf:28]: Same comment as above.

File: modules/azure_aks/main.tf [hunk 6]
```diff
@@ -1,4 +1,4 @@
-# Reference: https://www.terraform.io/docs/providers/azurerm/r/kubernetes_cluster.html
+# Reference:
```
Comment (thpang) [inline modules/azure_aks/main.tf:1]: Updated link: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster

**Review by thpang — CHANGES_REQUESTED** [review]
Looks good overall for the advanced networking, would like some logic changes and some of the changes look like they really need to be in another pr to address the added flexibility of those items and more documentation.

File: main.tf (hunk 1 above)
Comment (enderm) [reply main.tf:77]: @thpang just wondering why we would want to have a smaller subnet for "kubenet". We have the whole VPC at our disposal, and we are using only a tiny fraction of it. If anything, we should make the subnets larger to make use of the available ip range space so users don't run into issues if they happen to have a larger number of nodes, say.

File: docs/CONFIG-VARS.md (hunk 2 above)
Comment (enderm) [reply docs/CONFIG-VARS.md:105]: does enable_node_public_ip actually make sense? The nodes sit on a private subnet, so there is no way to route to their public ip's directly anyway.

File: variables.tf [hunk 7]
```diff
@@ -102,6 +97,52 @@ variable "default_nodepool_availability_zones" {
...
+variable "aks_service_cidr" {
+  description = "The Network Range used by the Kubernetes service. Changing this forces a new resource to be created."
+  default     = "10.0.0.0/16"
+}
+
+variable "node_vm_admin" {
+  description = "OS Admin User for VMs of AKS Cluster nodes"
+  default     = "azureuser"
+}
+
```
Comment (enderm) [inline variables.tf:145]: I'm not sure about the use case for surfacing the ip and cidr ranges as variables. They could also just be local variables.
I'm thinking about UI simplicity vs functionality.

File: main.tf (hunk 1 above)
Comment (thpang) [reply main.tf:77]: I can see your point. Currently the total hosts that can be on the AKS subnet is 256. I don't disagree there is always room for growth. With the new CIDR range the AKS subnet will have access to 4096 hosts. If we're going to adjust this value here we should also adjust the same values in the other providers. My suggestion would be to go with: 192.168.4.0/20 to keep things aligned. I am guessing we just took the 192.168.16.0/20 block from Barb's work.

File: docs/CONFIG-VARS.md (hunk 2 above)
Comment (thpang) [reply docs/CONFIG-VARS.md:105]: I was thinking the same thing, but again, believe these value additions should be in another pr.

File: main.tf (hunk 1 above)
Comment (manoatsas) [reply main.tf:77]: Using the CIDR range based on Azure error message. To @enderm comment it could be opened up to larger range.

Comment (enderm) [reply main.tf:77]: yeah, cidr range calculations are so not intuitive. I found this site helpful for avoiding range overlaps: http://www.davidc.net/sites/default/subnets/subnets.html

File: docs/CONFIG-VARS.md (hunk 2 above)
Comment (manoatsas) [reply docs/CONFIG-VARS.md:105]: These are to expose all available configs to the user, should they want to exercise any and all of these - https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster_node_pool. I could see cloud cost conscious user's wanting to take advantage of 'Spot' VM prices.

Have to check on 'enable_node_public_ip=true' with subnet config, could be removed.

File: modules/aks_node_pool/main.tf (hunk 3 above)
Comment (manoatsas) [reply modules/aks_node_pool/main.tf:12]: limiting the scope here to these changes not knowing all the changes that v2.38.0 or the latest v2.40.0 would require.

File: modules/aks_node_pool/main.tf (hunk 4 above)
Comment (manoatsas) [reply modules/aks_node_pool/main.tf:43]: same as above

File: modules/azure_aks/main.tf (hunk 6 above)
Comment (manoatsas) [reply modules/azure_aks/main.tf:1]: Redirects to the link

**Comment by NormJohnIV** [comment]
Since the system/default nodepool is already separate from the additional node_pool what is the reasoning behind allowing setting of the mode in the additional nodepools? They should always be User correct? This seems like it may be an unnecessary option.

File: modules/azure_aks/main.tf (hunk 6 above)
Comment (thpang) [reply modules/azure_aks/main.tf:1]: What is `Redirects to the link` ? There is no link to reference and/or redirect too. That's why I supplied the link. It was removed from the code.

File: main.tf (hunk 1 above)
Comment (thpang) [reply main.tf:77]: Ok, then best to keep the 192.168.16.0/20 block. Great tool @enderm Very helpful. Also might need text on this in the code and make sure we propagate this into the other providers.

File: docs/CONFIG-VARS.md (hunk 2 above)
Comment (thpang) [reply docs/CONFIG-VARS.md:105]: I agree, I just don't believe it's part of this pr. They might be needed and do provide flexibility, just not answer this pr's question. My vote would be to open another pr and present these changes there.

**Comment by manoatsas** [comment]
> Since the system/default nodepool is already separate from the additional node_pool what is the reasoning behind allowing setting of the mode in the additional nodepools? They should always be User correct? This seems like it may be an unnecessary option.

@NormJohnIV I initially thought so too but then some usecases like performance validation, HA testing, deploy Viya4 monitoring into a separate "system" nodepool or when 'System' nodepool created by default_nodepool should fail, there needs to be a way for users create a new 'System' nodepool to cordon off all system pods without taking down Viya4 deployments and/or entire AKS cluster. Without providing the 'mode' I don't see another way, agree?

File: modules/azure_aks/main.tf (hunk 6 above)
Comment (manoatsas) [reply modules/azure_aks/main.tf]: oops.. missed that the previous link was removed entirely, fixed it!

**Comment by thpang** [comment]
After reading over how `system` node pools work with Azure and based on information from this article : 

https://pumpingco.de/blog/modify-aks-default-node-pool-in-terraform-without-redeploying-the-cluster/

It's best we do not expose the `mode` flag at this time. This is something the kubernetes admin would be taking care of and would possibly do outside of terraform and the sync later. This addition would need more investigation.

@manoatsas if you can make that last change we can merge this pr. Thx.

**Review by thpang — COMMENTED** [review]
Changes look good.

**Review by thpang — APPROVED** [review]
